src/Compilers/Core/Portable/InternalUtilities/InterlockedOperations.cs

```diff
@@ -62,5 +63,29 @@ public static ImmutableArray<T> Initialize<T>(ref ImmutableArray<T> target, Immu
             var oldValue = ImmutableInterlocked.InterlockedCompareExchange(ref target, initializedValue, default(ImmutableArray<T>));
             return oldValue.IsDefault ? initializedValue : oldValue;
         }
+
+        /// <summary>
+        /// Initialize the immutable array referenced by <paramref name="target"/> in a thread-safe manner.
+        /// </summary>
+        /// <typeparam name="T">Elemental type of the array.</typeparam>
+        /// <param name="createArray">Callback to produce the array if <paramref name="target"/> is 'default'.  Will
+        /// only be called at most once.  Will not be called if 'target' is not 'default'.</param>
```

Comment:
> Will only be called at most once.

This is _absolutely_ not true. In the face of parallel initialization calls, it can invoke the callback multiple times. However, all calls will return the same result (one will win).

Reply:
ah, you're right.  will fix.

Reply:
💭 There are two ways to interpret this. On one hand, there is a question (1) of whether a single invocation of this method will call the method more than once. On the other hand, there is a question (2) of whether multiple invocations of `InterlockedInitialize` in parallel can cause one or more extraneous callback invocations that are eventually ignored.

In practice, I believe scenario (1) can only occur in a parallel initialization scenario where one of the initialization methods returned `default` and left the target in an uninitialized state. Scenario (2) is more likely to occur in parallel initialization cases.